ntp-os-clock/src/lib.rs

```diff
@@ -117,29 +122,74 @@ impl NtpClock for UnixNtpClock {
         ))
     }
 
-    fn adjust_clock(offset: ntp_proto::NtpDuration, freq_offset_ppm: f64) -> Result<(), Error> {
-        if freq_offset_ppm.abs() > 500.0 {
-            return Err(Error::Invalid);
-        }
-
+    fn set_freq(&self, freq: f64) -> Result<(), Self::Error> {
         let mut ntp_kapi_timex = EMPTY_TIMEX;
-        ntp_kapi_timex.modes = libc::ADJ_FREQUENCY | libc::ADJ_NANO;
-        ntp_kapi_timex.freq = (freq_offset_ppm * ((1 << 16) as f64)) as libc::c_long;
-
-        let (offset_seconds, offset_nanos) = offset.as_seconds_nanos();
-
-        // gradual adjustment if less than half a second off, otherwise jump
-        if (offset_seconds == -1 && offset_nanos >= 500_000_000)
-            || (offset_seconds == 0 && offset_nanos <= 500_000_000)
-        {
-            ntp_kapi_timex.modes |= libc::ADJ_OFFSET;
-            ntp_kapi_timex.offset =
-                (offset_nanos as i32 + offset_seconds * 1_000_000_000) as libc::c_long;
+        ntp_kapi_timex.modes = libc::MOD_FREQUENCY;
+        ntp_kapi_timex.freq = (freq * 65536e6) as libc::c_long;
+        if unsafe { libc::ntp_adjtime(&mut ntp_kapi_timex as *mut _) } != -1 {
+            // We don't care here about the time status, so the non-error
+            // information in the return value of ntp_adjtime can be ignored
+            Ok(())
         } else {
-            ntp_kapi_timex.modes |= libc::ADJ_SETOFFSET;
-            ntp_kapi_timex.time.tv_sec = offset_seconds as libc::time_t;
-            ntp_kapi_timex.time.tv_usec = offset_nanos as libc::suseconds_t;
+            Err(convert_errno())
+        }
+    }
+
+    fn step_clock(&self, offset: ntp_proto::NtpDuration) -> Result<(), Self::Error> {
+        let mut tp = libc::timespec {
+            tv_sec: 0,
+            tv_nsec: 0,
+        };
+
+        let (offset_secs, offset_nanos) = offset.as_seconds_nanos();
+
+        // Begin time critical section
+        // any time spend between here and the clock_settime call will reduce the
+        // accuracy of the made step
+        if unsafe { libc::clock_gettime(libc::CLOCK_REALTIME, &mut tp as *mut _) } == -1 {
+            return Err(convert_errno());
+        }
+
+        tp.tv_sec += offset_secs as libc::time_t;
+        tp.tv_nsec += offset_nanos as libc::c_long;
+        while tp.tv_nsec >= 1_000_000_000 {
+            tp.tv_nsec -= 1_000_000_000;
+            tp.tv_sec += 1;
+        }
```

Comment:
we expect this to usually do 0 or 1 iterations? when can this happen? maybe add a comment?

Reply:
because you can make this branchless but maybe there is no point?

Reply:
Anyway, made it an if to indicate intent more clearly

Reply:
unless I'm missing something
```
tp.tv_sec += ((tp.tv_nsec / 1_000_000_000) == 0) as _;
tp.tv_nsec %= 1_000_000_000;
```

now, I'm not sure it's actually better, but here it is

Reply:
I dont think it is better, for one thing, it uses modulo and division, which is slow. Anyway, hopefully clang is smart enough to recognize that it can convert the if into a number of cmovs in release builds, which should be the fastest solution.